File: frontend/src/main/java/app/programmatic/ui/reporting/domains/DomainsReportService.java
package app.programmatic.ui.reporting.domains;

import app.programmatic.ui.reporting.model.ReportFormat;
import app.programmatic.ui.reporting.validation.ValidateReportParameters;
import app.programmatic.ui.reporting.view.ReportMeta;

public interface DomainsReportService {
    byte[] runReportDomains(@ValidateReportParameters DomainsReportParameters parameters, ReportFormat format);

    ReportMeta getReportMeta(DomainsReportParameters parameters);
}

File: ui/src/app/advertiser/services/domains_report.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Router} from '@angular/router';
import {CommonService} from '../../common/services/common.service';

@Injectable()
export class DomainsReportService extends CommonService {

  constructor(public router: Router, public http: HttpClient) {
    super(router, http);
  }

  getReportMeta(parameters: DomainsReportParameters): Promise<ReportMeta> {
    return this.httpPut(this.api.report.meta.domains, parameters);
  }

  generateReport(parameters: DomainsReportParameters): Promise<any> {
    return this.httpPut(this.api.report.run.domains, parameters, {format: 'JSON'});
  }

  downloadReport(parameters: DomainsReportParameters, format: string): Promise<any> {
    return this.httpDownload('PUT', this.api.report.run.domains, parameters, {format});
  }
}
